### ISSUE TITLE
 Writing NN index file to temp file is slow

### ISSUE THREAD

**Issue by SansMorel** [jlmelville/uwot#27]
Why does writing NN index file to temp take so long? Is it possible to speed it up? 

`merged` is a large numeric matrix. 

Input: 
```
markers <- c(19:33,36:51,53,62)
sub <- merged[,]
library(uwot)
threads <- 32
umap_data <- umap(
  sub[,markers],
  n_neighbors = 15,
  n_components = 2,
  metric = "euclidean",
  n_epochs = 1000,
  learning_rate = 1,
  scale = "z",
  init = "spca",
  init_sdev = NULL,
  # spread = 5,
  min_dist = 0.01,
  set_op_mix_ratio = 1,
  local_connectivity = 1,
  bandwidth = 1,
  repulsion_strength = 1,
  negative_sample_rate = 5,
  nn_method = "annoy",
  # n_trees = 50,
  approx_pow = FALSE,
  pca = NULL,
  pca_center = TRUE,
  pcg_rand = TRUE,
  fast_sgd = FALSE,
  ret_model = FALSE,
  ret_nn = FALSE,
  n_threads = threads,
  n_sgd_threads = threads,
  grain_size = 1,
  verbose = TRUE
)
```

Output: 
> 19:37:09 Read 13869323 rows and found 33 numeric columns
> 19:37:09 Scaling to zero mean and unit variance
> 19:37:16 Kept 33 non-zero-variance columns
> 19:37:38 Building Annoy index with metric = euclidean, n_trees = 50
> 0%   10   20   30   40   50   60   70   80   90   100%
> [----|----|----|----|----|----|----|----|----|----|
> **************************************************|
> 20:29:09 Writing NN index file to temp file C:\Users\*\AppData\Local\Temp\Rtmp2D6YGc\file47d421456b98

It has been stuck on the last step for more than 15 hours. File size is about 1.8GB. 

With 1e5 rows: 
```
markers <- c(19:33,36:51,53,62)
sub <- merged[1:1e5,]
library(uwot)
threads = 64
umap_data <- umap(
  sub[,markers],
  n_neighbors = 15,
  n_components = 2,
  metric = "euclidean",
  n_epochs = 500,
  learning_rate = 1,
  scale = "z",
  init = "spca",
  init_sdev = NULL,
  # spread = 5,
  min_dist = 0.01,
  set_op_mix_ratio = 1,
  local_connectivity = 1,
  bandwidth = 1,
  repulsion_strength = 1,
  negative_sample_rate = 5,
  nn_method = "annoy",
  # n_trees = 50,
  approx_pow = FALSE,
  pca = NULL,
  pca_center = TRUE,
  pcg_rand = TRUE,
  fast_sgd = FALSE,
  ret_model = FALSE,
  ret_nn = FALSE,
  n_threads = threads,
  n_sgd_threads = threads,
  grain_size = 1,
  verbose = TRUE
)
```

> 11:57:16 Read 100000 rows and found 33 numeric columns
> 11:57:16 Scaling to zero mean and unit variance
> 11:57:16 Kept 33 non-zero-variance columns
> 11:57:17 Building Annoy index with metric = euclidean, n_trees = 50
> 0%   10   20   30   40   50   60   70   80   90   100%
> [----|----|----|----|----|----|----|----|----|----|
> **************************************************|
> 11:57:30 Writing NN index file to temp file C:\Users\*\AppData\Local\Temp\Rtmpcf6MHg\file1c05eb14ee7
> 11:57:30 Searching Annoy index using 64 threads, search_k = 1500
> 12:00:22 Annoy recall = 100%
> 12:00:23 Commencing smooth kNN distance calibration using 64 threads
> 12:00:25 Initializing from PCA
> 12:00:25 PCA: 2 components explained 27.42% variance
> 12:00:25 Commencing optimization for 500 epochs, with 2325534 positive edges using 64 threads
> 0%   10   20   30   40   50   60   70   80   90   100%
> [----|----|----|----|----|----|----|----|----|----|
> **************************************************|
> 12:00:53 Optimization finished

So with 1e5 rows the writing the NN index takes not even a second and file size is 73MB. 
With 1e6 rows: 

> 12:02:21 Read 1000000 rows and found 33 numeric columns
> 12:02:21 Scaling to zero mean and unit variance
> 12:02:22 Kept 33 non-zero-variance columns
> 12:02:23 Building Annoy index with metric = euclidean, n_trees = 50
> 0%   10   20   30   40   50   60   70   80   90   100%
> [----|----|----|----|----|----|----|----|----|----|
> **************************************************|
> 12:05:16 Writing NN index file to temp file C:\Users\*\AppData\Local\Temp\Rtmpcf6MHg\file1c01733238f
> 12:05:17 Searching Annoy index using 64 threads, search_k = 1500

With 1e6 rows, writing NN index takes 1 second and file size is about 738MB.

Specs:
Samsung EVO SSD 1 TB
128 GB ECC RAM
AMD 2990WX

`SessionInfo()`
> R version 3.6.0 (2019-04-26)
> Platform: x86_64-w64-mingw32/x64 (64-bit)
> Running under: Windows >= 8 x64 (build 9200)
> 
> Matrix products: default
> 
> locale:
> [1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
> [3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
> [5] LC_TIME=English_United Kingdom.1252    
> 
> attached base packages:
> [1] stats     graphics  grDevices utils     datasets  methods   base     
> 
> other attached packages:
> [1] uwot_0.1.3      Matrix_1.2-17   foreach_1.4.4   flowCore_1.50.0
> 
> loaded via a namespace (and not attached):
>  [1] graph_1.62.0        Rcpp_1.0.1          cluster_2.0.8       BiocGenerics_0.30.0
>  [5] MASS_7.3-51.4       lattice_0.20-38     rrcov_1.4-7         pcaPP_1.9-73       
>  [9] vizier_0.3          tools_3.6.0         parallel_3.6.0      grid_3.6.0         
> [13] Biobase_2.44.0      snow_0.4-3          corpcor_1.6.9       iterators_1.0.10   
> [17] matrixStats_0.54.0  RcppParallel_4.4.3  doSNOW_1.0.16       codetools_0.2-16   
> [21] robustbase_0.93-5   compiler_3.6.0      DEoptimR_1.0-8      stats4_3.6.0       
> [25] mvtnorm_1.0-10

**Comment by SansMorel** [comment]
[https://github.com/jlmelville/uwot/issues/26](url)

> Annoy will fail on large datasets due to not being able to read back in an index larger in size than 2GB

**Comment by jlmelville** [comment]
I've not seen this exact problem before, but it is due to the index being too large to be written to disk. I'm sorry that it is causing a hang and not failing early. Unfortunately, I don't think I can detect whether this is going to be a problem without access to Annoy's internals, which would require upstream fixes to Annoy and then RcppAnnoy.

This issue is a large part of why I would like to move to using the rp-forest/nn-descent method for nearest neighbor search used by the Python UMAP. I'm sorry that there is no immediate fix for this. Only using one thread will eventually work, but that's obviously not a very desirable state of affairs for large datasets on a multi-core machine.
